components/places/src/db/interrupt.rs

```diff
@@ -0,0 +1,77 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use crate::error::*;
+use rusqlite::InterruptHandle;
+use std::sync::{
+    atomic::{AtomicUsize, Ordering},
+    Arc,
+};
+
+// SeqCst is overkill for much of this, but whatever.
+
+/// A Sync+Send type which can be used allow someone to interrupt an
+/// operation, even if it happens while rust code (and not SQL) is
+/// executing.
+pub struct PlacesInterruptHandle {
+    pub(crate) db_handle: InterruptHandle,
+    pub(crate) interrupt_counter: Arc<AtomicUsize>,
+}
+
+impl PlacesInterruptHandle {
+    pub fn interrupt(&self) {
+        self.interrupt_counter.fetch_add(1, Ordering::SeqCst);
+        self.db_handle.interrupt();
+    }
+}
+
+/// A helper that can be used to determine if an interrupt request has come in while
+/// the object lives. This is used to avoid a case where we aren't running any
+/// queries when the request to stop comes in, but we're still not done (for example,
+/// maybe we've run some of the autocomplete matchers, and are about to start
+/// running the others. If we rely solely on sqlite3_interrupt(), we'd miss
+/// the message that we should stop).
+pub(crate) struct InterruptScope {
```

Comment:
Oh, this is clever; it's a [cancellation token](https://docs.microsoft.com/en-us/dotnet/api/system.threading.cancellationtoken)! This might be premature generalization, and definitely not in this PR, but worth thinking about factoring this out into a more generic cooperative cancellation mechanism that we can use elsewhere in a-s.